extensions/websockets-next/deployment/src/test/java/io/quarkus/websockets/next/test/security/HttpUpgradePermissionsAllowedAnnotationTest.java

```diff
@@ -0,0 +1,142 @@
+                            AdminEndpoint.class, InclusiveEndpoint.class));
+
+    @TestHTTPResource("admin-end")
+    URI adminEndpointUri;
+
+    @TestHTTPResource("inclusive-end")
+    URI inclusiveEndpointUri;
+
+    @Test
+    public void testInsufficientRights() {
+        try (WSClient client = new WSClient(vertx)) {
+            CompletionException ce = assertThrows(CompletionException.class,
+                    () -> client.connect(basicAuth("user", "user"), adminEndpointUri));
+            Throwable root = ExceptionUtil.getRootCause(ce);
+            assertInstanceOf(UpgradeRejectedException.class, root);
+            assertTrue(root.getMessage().contains("403"));
+        }
+        try (WSClient client = new WSClient(vertx)) {
+            client.connect(basicAuth("admin", "admin"), adminEndpointUri);
+            client.waitForMessages(1);
+            assertEquals("ready", client.getMessages().get(0).toString());
+            client.sendAndAwait("hello");
+            client.waitForMessages(2);
+            assertEquals("hello", client.getMessages().get(1).toString());
+        }
+    }
+
+    @Test
+    public void testInclusivePermissions() {
+        Stream.of("admin", "user").forEach(name -> {
+            try (WSClient client = new WSClient(vertx)) {
+                CompletionException ce = assertThrows(CompletionException.class,
+                        () -> client.connect(basicAuth(name, name), inclusiveEndpointUri));
+                Throwable root = ExceptionUtil.getRootCause(ce);
+                assertInstanceOf(UpgradeRejectedException.class, root);
+                assertTrue(root.getMessage().contains("403"));
+            }
+        });
+        try (WSClient client = new WSClient(vertx)) {
+            client.connect(basicAuth("almighty", "almighty"), inclusiveEndpointUri);
+            client.waitForMessages(1);
+            assertEquals("ready", client.getMessages().get(0).toString());
+            client.sendAndAwait("hello");
+            client.waitForMessages(2);
+            assertEquals("hello", client.getMessages().get(1).toString());
+        }
+    }
+
+    @PermissionsAllowed(value = { "perm1", "perm2" }, inclusive = true)
+    @WebSocket(path = "/inclusive-end")
+    public static class InclusiveEndpoint {
+
+        @OnOpen
+        String open() {
+            return "ready";
+        }
+
+        @OnTextMessage
+        String echo(String message) {
+            return message;
+        }
+
+    }
+
+    @PermissionsAllowed("endpoint:read")
+    @WebSocket(path = "/admin-end")
+    public static class AdminEndpoint {
+
+        @OnOpen
+        String open() {
+            return "ready";
+        }
+
+        @OnTextMessage
+        String echo(String message) {
+            return message;
+        }
+
+    }
+
+    @PermissionsAllowed(value = { "endpoint:connect", "endpoint:read" })
+    @WebSocket(path = "/end")
+    public static class Endpoint {
```

Comment:
Is it used here?

Reply:
yes, it is tested by SecurityTestBase and unit tests declared directly in this test class are just complementing the base tests

Reply:
please see `extends SecurityTestBase`